runtime/src/main/cpp/Memory.cpp

```diff
@@ -1125,15 +1125,23 @@ inline size_t containerSize(const ContainerHeader* container) {
 }  // namespace
 
 MetaObjHeader* ObjHeader::createMetaObject(TypeInfo** location) {
-  MetaObjHeader* meta = konanConstructInstance<MetaObjHeader>();
   TypeInfo* typeInfo = *location;
   RuntimeCheck(!hasPointerBits(typeInfo, OBJECT_TAG_MASK), "Object must not be tagged");
+
+#if !KONAN_NO_THREADS
+  if (typeInfo->typeInfo_ != typeInfo) {
+    // Someone installed a new meta-object since the check.
+    return reinterpret_cast<MetaObjHeader*>(typeInfo);
+  }
+#endif
+
+  MetaObjHeader* meta = konanConstructInstance<MetaObjHeader>();
   meta->typeInfo_ = typeInfo;
 #if KONAN_NO_THREADS
   *location = reinterpret_cast<TypeInfo*>(meta);
 #else
   TypeInfo* old = __sync_val_compare_and_swap(location, typeInfo, reinterpret_cast<TypeInfo*>(meta));
-  if (old->typeInfo_ != old) {
+  if (old != typeInfo) {
```

Comment:
In my tests just this line is sufficient for race to go away, code above just duplicates the check in `has_meta_object()`. My test was was Swift program from my comment in #2931 with 100 background jobs enclosed in while loop and with another commit from this PR applied.

Reply:
In my tests just this line is not sufficient for race to go away.

`test.kt`:
```kotlin
enum class Direction {
    NORTH, SOUTH, WEST, EAST, NORTHWEST, NORTHEAST
}

fun toString(obj: Any) = obj.toString()
```

`test.swift`:
```swift
import Test

func loadInBackground()  {
    DispatchQueue.global(qos: .background).async {
        TestKt.toString(obj: Direction.east)
    }
}

for i in 0..<2 {
    loadInBackground()
}

usleep(100000)
```

> code above just duplicates the check in `has_meta_object()`

Code above doesn't just duplicate the check in `has_meta_object()`. It prevents from data race when `typeInfo` is different from that read in `has_meta_object()`:

1.  Thread 1 loads original type info and ensures that `has_meta_object()` is `false`. Then it allocates `meta1`.
2.  Thread 2 successfully checks, allocates and stores meta object `meta2`.
3.  Thread 1 loads `meta2` as `typeInfo` and makes `meta1` pointing `meta2` instead of original type info. Then it makes successful compare-and-swap, and now the object points to `meta1` which points to `meta2`.